### DISCUSSION TITLE
When will ccache 4.0 be released?

### DISCUSSION THREAD

**Discussion by dyadav7** [ccache/ccache#653] [Questions & answers] [answered]
### Question ###

i want to use the latest ccache from the master version. I would like an official release to install.
When will there be an official release from the master branch ?

**Comment by jrosdahl** [comment, marked as answer]
That's a reasonable question, which unfortunately doesn't have a good answer. Like many open source projects, ccache is driven by volunteers who do what they like to do when they have time to do it on their spare time.

The most accurate answer is: There is no time plan. It will be released when it's ready.

What's left before it's ready, you then ask? Well, since 4.0 is a major ("breaking") version bump, I want to include the items on my to do list that are backward incompatible in one way or another. Many of those things have now been implemented (e.g. aggregated result files, new compression algorithm (Zstandard) enabled by default, new hash algorithm (BLAKE3), new tooling (C++, CMake), using $XDG_*_HOME directories by default). The only major thing left on my list now is revising the cache directory structure (in a mostly but not 100% backward compatible way). I don't keep my to do list in public because it would take too much time to maintain, but you can see a partial view of it by listing [open issues with milestone 4.0](https://github.com/ccache/ccache/milestones/4.0).

When I started the work on 4.0 last summer it felt like a release maybe could be about three months away. I've had that feeling since then. Classic. :slightly_smiling_face: In particular the change from C to C++ has been a honeypot for well-meaning contributors (and partially myself) who find pleasure in various refactoring and improvements. That's mostly good, but it also limits my available time for driving real features forward.

So here's another answer: The release may be about three months away.

**Comment by dyadav7** [comment]
Thanks for sharing so many details.
I will use the ccache from github directly for next 3 months,

**Comment by afbjorklund** [comment]
> In particular the change from C to C++ has been a honeypot for well-meaning contributors (and partially myself) who find pleasure in various refactoring and improvements

This was unfortunately expected, but hopefully it will improve the end result

**Comment by jrosdahl** [comment]
> The only major thing left on my list now is revising the cache directory structure

This has now been done so a release should not be many weeks away. Currently left to do is #432, writing release notes and waiting a while for people to try out the latest changes, primarily [automatic cache levels](https://github.com/ccache/ccache/commit/fd8474e3bd1463fdea8e95c40a2210fdcc93ed31) and [statistics on level 2](https://github.com/ccache/ccache/commit/fb7c03276ebd7201929f392149e974009dc08102).

**Comment by Romain-Geissler-1A** [comment]
Hi @jrosdahl,

I don't know if this is the right place to ask such questions, but I have a question about the upcoming ccache 4.0 release. Let's say we are in a company environment with thousands of developers, and many of them do share some common ccache folder (not every team uses the same cache directory, but it's not rare that a cache directory is shared by 10+ developers, sometimes 100+). These ccache directories are shared across different machines using NFS.

So far I always followed all ccache minor releases, and except one bug once, I think most of the time it went well. Now comes ccache 4.0, and the new directory structure. Given that not all users don't all move to the new ccache release all at the same time, yet they might share the same ccache directory, shall I be cautious in this migration ? In other words, is it safe to share the same ccache directory with a user using ccache 3.7.12 and a user using ccache 4.0 ?

Thanks,
Romain

**Comment by jrosdahl** [comment]
>  is it safe to share the same ccache directory with a user using ccache 3.7.12 and a user using ccache 4.0 ?

Yes, the cache structure has been modified in such a way that it should be safe to run pre-4.0 and post-4.0 versions simultaneously against the same cache directory. I will write something about this in the release notes.

**Comment by jrosdahl** [comment]
4.0 has now been released.
